File: leetcode-rs/1300_1399/leetcode-no1385.rs
// problem: https://leetcode.cn/problems/find-the-distance-value-between-two-arrays/
struct Solution;

use std::cmp::min;
impl Solution {
    pub fn find_the_distance_value(mut arr1: Vec<i32>, mut arr2: Vec<i32>, d: i32) -> i32 {
        let base = 1000000;
        let limit = base - 2000;
        arr1.extend(arr2.into_iter().map(|num| num + base));
        arr1.sort_by(|mut num1, mut num2| {
            let mut n1 = *num1;
            let mut n2 = *num2;
            if n1 > limit {
                n1 -= base;
            }
            if n2 > limit {
                n2 -= base;
            }
            return n1.cmp(&n2);
        });
        let mut res = i32::MAX;
        let n = arr1.len();
        for i in 0..n - 1 {
            let first = arr1[i];
            let second = arr1[i + 1];
            let mut f_is = false;
            let mut s_is = false;
            if first > limit {
                f_is = true;
            }
            if second > limit {
                s_is = true;
            }
            if f_is ^ s_is {
                res = min(res, (arr1[i] - arr1[i - 1]).abs());
            }
        }
        res
    }
}
fn main() {
    println!(
        "{:?} e:2",
        Solution::find_the_distance_value(vec![4, 5, 8], vec![10, 9, 1, 8],2)
    );
    println!(
        "{:?} e:2",
        Solution::find_the_distance_value(vec![1,3,2,3],vec![-4,-3,6,10,20,30],2)
    );
    println!(
        "{:?} e:6",
        Solution::find_the_distance_value(vec![2,1,100,3],vec![-5,-2,10,-3,7],2)
    );
}

//
// Created by jing at 2025-09-07 15:47
//

File: leetcode-cpp/1300_1399/leetcode-no1385.cpp
// problem:
// https://leetcode.cn/problems/find-the-distance-value-between-two-arrays/?envType=problem-list-v2&envId=FeuPdeiM
#include <algorithm>
#include <iostream>
#include <ranges>
#include <vector>
#include "../utils/vector_helper.h"


using namespace std;

class Solution2 {
public:
    int findTheDistanceValue(vector<int> &arr1, vector<int> &arr2, int d) {
        ranges::sort(arr2);
        int n = arr1.size(), res = 0;
        for (int i = 0; i < n; ++i) {
            auto closest = lower_bound(arr2.begin(), arr2.end(), arr1[i]);
            int dist = INT_MAX;
            if (closest != arr2.end()) {
                dist = min(dist, *closest - arr1[i]);
            }
            if (closest != arr2.begin()) {
                dist = min(dist, arr1[i] - *(closest - 1));
            }
            if (dist > d) res++;
        }
        return res;
    }
};

class Solution {
public:
    int findTheDistanceValue(vector<int> &arr1, vector<int> &arr2, int d) {
        ranges::sort(arr2), ranges::sort(arr1);
        int n = arr1.size(), m = arr2.size(), index = 0, res = 0;
        for (const int &num: arr1) {
            while (index < m and arr2[index] <= num) index++;
            int dist = INT_MAX;
            if (index > 0) dist = num - arr2[index - 1];
            if (index < m) dist = min(dist, arr2[index] - num);
            if (dist > d) res++;
        }
        return res;
    }
};

int main(int argc, char *argv[]) {
    Solution test;
    cout << test.findTheDistanceValue(temp_vector({4, 5, 8}), temp_vector({10, 9, 1, 8}), 2);
    cout << test.findTheDistanceValue(temp_vector({1, 4, 2, 3}), temp_vector({-4, -3, 6, 10, 20, 30}), 3);
}

//
// Created By jing At 2026-05-25 18:17
//
